### ISSUE TITLE
Fallback plugin logic faulty

### ISSUE THREAD

**Issue by jni** [napari/napari#1213]
## 🐛 Bug

Dragging and dropping an (apparently malformed 😅) multiscale .zarr folder to napari gave me the following error:

```pytb
INFO:napari_plugin_engine.manager:loaded 3 plugins:
  <module 'napari.plugins._builtins' from '/Users/jni/conda/envs/all/lib/python3.8/site-packages/napari/plugins/_builtins.py'>
  <module 'napari_svg' from '/Users/jni/conda/envs/all/lib/python3.8/site-packages/napari_svg/__init__.py'>
  <module 'ome_zarr' from '/Users/jni/conda/envs/all/lib/python3.8/site-packages/ome_zarr.py'>
  <module 'napari_plugin_engine' from '/Users/jni/conda/envs/all/lib/python3.8/site-packages/napari_plugin_engine/__init__.py'>
ERROR:napari.plugins.io:PluginError: Error in plugin 'builtins', hook 'napari_get_reader': local variable 'image' referenced before assignment
  Cause was: local variable 'image' referenced before assignment
    version: 0.3.0
    package: napari

ERROR:napari.components.add_layers_mixin:No plugin found capable of reading /Volumes/datac/napari-data/covid-gut/.
```

Note "local variable image referenced before assignment". This should never happen.

**Comment by jni** [comment]
Another logic issue: if `data` is a list BUT the arrays are *not* in decreasing order of size, we throw an AttributeError:

```pytb
~/conda/envs/all/lib/python3.8/site-packages/napari/layers/image/image.py in __init__(self, data, rgb, colormap, contrast_limits, gamma, interpolation, rendering, iso_threshold, attenuation, name, metadata, scale, translate, opacity, blending, visible, multiscale)
    162             init_shape = data[0].shape
    163         else:
--> 164             init_shape = data.shape
    165
    166         # Determine if rgb

AttributeError: 'list' object has no attribute 'shape'
```

In this case this happened because the plugin (I'm still trying to figure out why) ended up providing the arrays in lexicographical (0, 1, 10, 2) order rather than numerical.

We should detect malformed data and quit gracefully (or re-sort, but I like that less), rather than crash with a mysterious error.

**Comment by tlambert03** [comment]
do you have the full stack trace on that first one?   I _think_ (but I'm not sure) that that's an error in `magic_imread` (based on the variable name `image`) ...
(plugin errors will usually be using `raise from` ... so the plugin error is probably the one above that one)

**Comment by jni** [comment]
@tlambert03 

> do you have the full stack trace

Not just yet, and that might be a problem. ;) ie I can't get the stack trace if I launched napari from the command line.

Anyway, I think you're right. In this chunk:

```python
    for filename in filenames_expanded:
        ext = os.path.splitext(filename)[-1]
        if ext == '.zarr':
            image, zarr_shape = read_zarr_dataset(filename)
            if shape is None:
                shape = zarr_shape
        else:
            if shape is None:
                image = imread(filename)
                shape = image.shape
                dtype = image.dtype
            if use_dask:
                image = da.from_delayed(
                    delayed(imread)(filename), shape=shape, dtype=dtype
                )
            elif len(images) > 0:  # not read by shape clause
                image = imread(filename)
        images.append(image)
```

image can be unassigned by the time you hit append.

I'll keep investigating.

**Comment by tlambert03** [comment]
> I can't get the stack trace if I launched napari from the command line.

not even in the plugin error reporter?  under builtins?

ohhh ... do you mean you launch it with a filepath and it quits immediately??  hadn't thought about that!

**Comment by tlambert03** [comment]
if it's the covid-gut folder you're using, from Josh's download code the other day...  I get the same error.  i just ran it through a debugger and it's a couple things:
1) `filenames_expanded` remains empty after the `if os.path.isdir(filename) and not ext == '.zarr':` block... because it strips all the subdirectories...
then
2) line 224 then fails because `image = np.stack([])` is a `ValueError` ... but unfortunately the attempt to catch the ValueError and provide a helpful message is (my) bad.  the exception clause should have a `else: raise e` to re-reraise the error regardless

... but while we're at it, we should probably add a ` if not images` clause to the beginning of that last block to provide a useful message

**Comment by jni** [comment]
> do you mean you launch it with a filepath and it quits immediately?? hadn't thought about that!

Well, the thing is, that should never ever ever happen, so it's fair enough. We just need to iron out all the bugs from failures when the plugins fail.

**Comment by jni** [comment]
btw, @tlambert03, yes, it's from that covid folder, except you need to add a `.zgroup` file to the root folder with the contents `{"zarr_format":2}`.

After I do that, I get the second error due to the pyramid checks failing. Not sure yet if it's an issue with the plugin or with how I downloaded the folder. But if I delete the `10.zarr` level (and the corresponding dict in the json) it works! 🎉 😂

**Comment by tlambert03** [comment]
Awesome! Ok, well I’ll make a PR to fix those issues anyway, since they’re still bugs

**Comment by sofroniewn** [comment]
Closing this as main bugs were fixed by #1218 and remaining error now has its own issue in #1231
